nova_vm/src/ecmascript/builtins/structured_data/data_view_objects/data_view_constructor.rs

```diff
@@ -24,14 +37,136 @@ impl BuiltinIntrinsicConstructor for DataViewConstructor {
+            if buffer_is_fixed_length {
+                // i. Let viewByteLength be bufferByteLength - offset.
+                Some(buffer_byte_length - offset)
+            } else {
+                // b. Else,
+                // i. Let viewByteLength be auto.
+                None
+            }
+        } else {
+            // 9. Else,
+            // a. Let viewByteLength be ? ToIndex(byteLength).
+            let view_byte_length = to_index(agent, byte_length)? as usize;
+            // b. If offset + viewByteLength > bufferByteLength, throw a RangeError exception.
+            if offset + view_byte_length > buffer_byte_length {
+                return Err(agent.throw_exception_with_static_message(
+                    ExceptionType::RangeError,
+                    "offset is outside the bounds of the buffer",
+                ));
+            }
+            Some(view_byte_length)
+        };
+
+        // 10. Let O be ? OrdinaryCreateFromConstructor(NewTarget, "%DataView.prototype%", « [[DataView]], [[ViewedArrayBuffer]], [[ByteLength]], [[ByteOffset]] »).
+        let o = ordinary_create_from_constructor(agent, new_target, ProtoIntrinsics::DataView)?;
+
+        // 11. If IsDetachedBuffer(buffer) is true, throw a TypeError exception.
+        if is_detached_buffer(agent, buffer) {
+            return Err(agent.throw_exception_with_static_message(
+                ExceptionType::TypeError,
+                "attempting to access detached ArrayBuffer",
+            ));
+        }
+
+        // 12. Set bufferByteLength to ArrayBufferByteLength(buffer, seq-cst).
+        let buffer_byte_length = array_buffer_byte_length(agent, buffer, Ordering::SeqCst) as usize;
+
+        // 13. If offset > bufferByteLength, throw a RangeError exception.
+        if offset > buffer_byte_length {
+            return Err(agent.throw_exception_with_static_message(
+                ExceptionType::RangeError,
+                "offset is outside the bounds of the buffer",
+            ));
+        }
+
+        // 14. If byteLength is not undefined, then
+        if !byte_length.is_undefined() {
+            // a. If offset + viewByteLength > bufferByteLength, throw a RangeError exception.
+            if offset + view_byte_length.unwrap() > buffer_byte_length {
+                return Err(agent.throw_exception_with_static_message(
+                    ExceptionType::RangeError,
+                    "offset is outside the bounds of the buffer",
+                ));
+            }
+        }
+
+        let o = DataView::try_from(o).unwrap();
+
+        // 15. Set O.[[ViewedArrayBuffer]] to buffer.
+        agent[o].viewed_array_buffer = buffer;
+        // 16. Set O.[[ByteLength]] to viewByteLength.
+        agent[o].byte_length = view_byte_length.into();
+        // 17. Set O.[[ByteOffset]] to offset.
+        agent[o].byte_offset = offset.into();
+
+        if agent[o].byte_length == DataViewByteLength::heap() {
```

Comment:
**issue:** Here and below, don't re-read from the heap the value you just wrote into the heap but instead just use the on-stack value that you wrote there, so `view_byte_length` and `offset`.